driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -94,27 +94,43 @@ synchronized void rebuildSchema(String keyspace, String table, ResultSet ks, Res
                 KeyspaceMetadata ksm = KeyspaceMetadata.build(ksRow);
 
                 if (cfDefs.containsKey(ksName)) {
-                    buildTableMetadata(ksm, cfDefs.get(ksName), colsDefs.get(ksName), cassandraVersion);
+                    buildTableMetadata(ksm, null, cfDefs.get(ksName), colsDefs.get(ksName), cassandraVersion);
                 }
                 addedKs.add(ksName);
-                keyspaces.put(ksName, ksm);
+
+                KeyspaceMetadata previousKsm = keyspaces.get(ksName);
+                if (previousKsm == null) {
+                    keyspaces.put(ksName, ksm);
+                    triggerOnKeyspaceAdded(ksm);
+                } else if (!previousKsm.equals(ksm)) {
+                    keyspaces.put(ksName, ksm);
+                    triggerOnKeyspaceChanged(ksm, previousKsm);
+                }
             }
 
-            // If keyspace is null, it means we're rebuilding from scratch, so
-            // remove anything that was not just added as it means it's a dropped keyspace
-            if (keyspace == null) {
-                Iterator<String> iter = keyspaces.keySet().iterator();
-                while (iter.hasNext()) {
-                    if (!addedKs.contains(iter.next()))
-                        iter.remove();
+            Set<String> removedKs = new HashSet<String>();
+            for (String ksName : keyspaces.keySet()) {
```

Comment:
> we also need to broadcast dropped events for all tables in the dropped keyspace as well

Do we? I think it's implicit that if a keyspace got dropped, everything in it got dropped.

Apart from that, couldn't we do without the `removedKs` set?